### PR TITLE
[Fix] MoE DP reduce-scatter and logical axis sharding in vLLM serving path

### PR DESCRIPTION
**Pull request by sierraisland**
# Description

This PR closes the ~30% end-to-end serving performance gap between `MaxTextForCausalLM` and native vLLM (`MODEL_IMPL_TYPE=vllm`) when serving MoE models (e.g. `Qwen3.5-35B-A3B`) with DP Attention and Expert Parallelism enabled (`enable_dp_attention=True`, `expert_parallelism=8`, `tensor_parallelism=1`).

### Root Causes & Changes
1. **Pass `scatter_results` to `fused_moe_func` in `RoutedMoE.fused_moe_matmul` (`src/maxtext/layers/moe.py`)**:
   - In `tpu-inference` (`tpu_inference/layers/vllm/interface/moe.py`), `fused_moe_func` is invoked with `scatter_results=is_attn_dp(mesh)` (`True` when DP attention is enabled). This instructs `moe_gmm_local` to perform a 2-way `psum` over `expert` followed by a `psum_scatter` (`reduce-scatter`) across the DP attention axis (`attn_dp` / `attn_dp_expert`), directly returning hidden states sharded on `P(ShardingAxisName.ATTN_DATA)` (`[num_tokens // dp_size, hidden_size]` per DP rank) and enabling multi-chunk communication-computation pipelining (`actual_chunk_size = moe_chunk_size * scatter_axis_size`).
   - Previously, `RoutedMoE.fused_moe_matmul` omitted `scatter_results` (defaulting to `False`), causing every MoE layer to execute a full 8-chip `all-reduce` (`jax.lax.psum`) over the entire `[num_tokens, hidden_size]` buffer (`33.55 MB` per layer at `8,192` tokens) and return an unsharded `P('data')` tensor that then had to be sliced back down to `P(ATTN_DATA)`.
2. **Gate `self.selected_experts` computation behind `return_routed_experts` (`src/maxtext/layers/moe.py`)**:
   - `RoutedMoE.fused_moe_matmul` previously computed a duplicate `jax.lax.top_k(gating_output, self.num_experts_per_tok)` and stored `self.selected_experts = nnx.Intermediate(top_k_indices)` unconditionally on every layer (even though `fused_moe_func` already computes `top_k` internally), which `Decoder.__call__` concatenated and returned from the JIT model on every step. Gating this behind `getattr(self.config, "return_routed_experts", False)` avoids redundant `top_k` and concatenation during standard serving.
3. **Add `'expert'` axis to `mlp`, `activation_mlp`, and `vocab` in `src/maxtext/configs/inference/vllm.yml`**:
   - When running with `tensor_parallelism=1, expert_parallelism=8` (`Mesh: attn_dp=1, attn_dp_expert=4, expert=2, model=1`), attention rules (`heads`, `kv_heads`, `gdn_head`, `activation_vocab`) map to `['model', 'expert']` (2-way sharded across `expert=2`), whereas `mlp` and `activation_mlp` previously mapped only to `['attn_dp', 'model']` (`1 * 1 = 1`) and `vocab` mapped to `[]`.
   - This left Qwen3.5 full-attention `out_proj` (`out_kernel_axis=("mlp", "embed_attn")`), all shared expert MLP weights (`shared_experts`), and the vocabulary projection weights completely unsharded (`P(None, None)`). Adding `'expert'` aligns their 2-way tensor sharding with the rest of the model.

BUGS: b/549307194

# Tests

End to End Benchmark (`Qwen/Qwen3.5-35B-A3B` on TPU v7x-8, Multi-Turn Agentic GRPO Benchmark)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by khatwanimohit — APPROVED** [review]

**Review by A9isha — APPROVED** [review]

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -3645,8 +3446,9 @@ def fused_moe_matmul(
     hidden_states = jnp.reshape(inputs, (batch_size * seq_len, emb_dim))
     gating_output = jnp.reshape(gate_logits, (batch_size * seq_len, self.num_experts))
 
-    _, top_k_indices = jax.lax.top_k(gating_output, self.num_experts_per_tok)
-    self.selected_experts = nnx.Intermediate(top_k_indices)
+    if getattr(self.config, "return_routed_experts", False):
```
Comment (khatwanimohit) [inline src/maxtext/layers/moe.py]: we can remove getattr since it'll always be present in the config

File: src/maxtext/configs/base.yml [hunk 2]
```diff
@@ -317,6 +298,11 @@ use_gmm_v2_heuristic_tiling: false
 
 norm_topk_prob: false # boolean to enable the top-k probability normalization. qwen3-specific normalization of router weights.
 
+# Record each MoE layer's routing decisions as a `selected_experts` intermediate, which the decoder
+# returns alongside the hidden state as `expert_indices`. Needed to capture routing from inference
+# for router replay; off by default so standard serving does not pay for the extra top-k.
+return_routed_experts: false
```
Comment (khatwanimohit) [inline src/maxtext/configs/base.yml:335]: can you also add this option in rl.yml

File: src/maxtext/layers/moe.py [hunk 3]
```diff
@@ -3649,8 +3649,9 @@ def fused_moe_matmul(
     hidden_states = jnp.reshape(inputs, (batch_size * seq_len, emb_dim))
     gating_output = jnp.reshape(gate_logits, (batch_size * seq_len, self.num_experts))
 
-    _, top_k_indices = jax.lax.top_k(gating_output, self.num_experts_per_tok)
-    self.selected_experts = nnx.Intermediate(top_k_indices)
+    if self.config.return_routed_experts:
```
Comment (RissyRan) [inline src/maxtext/layers/moe.py:3652]: The default in base.yml as false, will this affect any training workload?

Comment (wenxindongwork) [reply src/maxtext/layers/moe.py:3652]: I don't think so, since training workloads don't need router replay.

Comment (RissyRan) [reply src/maxtext/layers/moe.py:3652]: Sounds good, and noticed that fused version only used for vLLM. https://screenshot-v2.corp.google.com/4iuibtbqo1ls0

**Review by RissyRan — APPROVED** [review]

File: src/maxtext/configs/base.yml [hunk 4]
```diff
@@ -327,6 +327,11 @@ use_gmm_v2_heuristic_tiling: false
 
 norm_topk_prob: false # boolean to enable the top-k probability normalization. qwen3-specific normalization of router weights.
 
+# Record each MoE layer's routing decisions as a `selected_experts` intermediate, which the decoder
```
Comment (RissyRan) [inline src/maxtext/configs/base.yml:330]: Will be great to note that this only used with vLLM in RL.

**Review by igorts-git — APPROVED** [review]
